inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_gather_nd.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "vpu/ngraph/transformations/dynamic_to_static_shape_gather.hpp"
+
+#include "vpu/ngraph/operations/dynamic_shape_resolver.hpp"
+#include "vpu/ngraph/utilities.hpp"
+#include <vpu/utils/error.hpp>
+#include "ngraph/graph_util.hpp"
+
+#include "ngraph/opsets/opset5.hpp"
+
+#include <numeric>
+
+namespace vpu {
+
+void dynamicToStaticShapeGatherND(std::shared_ptr<ngraph::Node> target) {
+    const auto gatherND = ngraph::as_type_ptr<ngraph::opset5::GatherND>(target);
+    VPU_THROW_UNLESS(gatherND, "dynamicToStaticShapeGatherND transformation is not applicable for {}, it should be {} instead",
+                     target, ngraph::opset5::GatherND::type_info);
+
+    auto shapeToConstant = [&gatherND](const ngraph::Output<ngraph::Node>& output,
+                                       const ngraph::element::Type& elemType) -> std::shared_ptr<ngraph::opset5::Constant> {
+        VPU_THROW_UNLESS(output.get_partial_shape().is_static(),
+                         "dynamicToStaticShapeGatherND transformation for {} of type {} expects static shape on inputs without DSR",
+                         gatherND->get_friendly_name(), gatherND->get_type_info());
+        return ngraph::opset5::Constant::create(elemType, {output.get_shape().size()}, output.get_shape());
+    };
+
+    const auto dataDSR = ngraph::as_type_ptr<ngraph::vpu::op::DynamicShapeResolver>(gatherND->input_value(0).get_node_shared_ptr());
+    const auto indicesDSR = ngraph::as_type_ptr<ngraph::vpu::op::DynamicShapeResolver>(gatherND->input_value(1).get_node_shared_ptr());
+
+    VPU_THROW_UNLESS(dataDSR || indicesDSR, "dynamicToStaticShapeGatherND transformation for {} of type {} expects at least one DSR as input",
+                     gatherND->get_friendly_name(), gatherND->get_type_info());
+    if (dataDSR && indicesDSR) {
+        VPU_THROW_UNLESS(dataDSR->get_input_element_type(1) == indicesDSR->get_input_element_type(1),
+                         "dynamicToStaticShapeGatherND transformation for {} of type {} expects equal shapes data types, actual {} vs {}",
+                         gatherND->get_friendly_name(), gatherND->get_type_info(),
+                         dataDSR->get_input_element_type(1), indicesDSR->get_input_element_type(1));
+    }
+    const auto shapeElementType = indicesDSR ? indicesDSR->get_input_element_type(1) : dataDSR->get_input_element_type(1);
+
+    const auto dataShape = dataDSR ? dataDSR->input_value(1) : shapeToConstant(gatherND->input_value(0), shapeElementType);
+    const auto indicesShape = indicesDSR ? indicesDSR->input_value(1) : shapeToConstant(gatherND->input_value(1), shapeElementType);
+
+    const auto dataShapeRank = static_cast<size_t>(dataShape.get_shape()[0]);
+    const auto indicesShapeRank = static_cast<size_t>(indicesShape.get_shape()[0]);
```

Comment:
Just because the number of elements in shape is a rank of data, and it's always 1-dimensional. I've replaced it with `shape_size`, to be more precise.

Reply:
Rank is number of dimensions in shape, not number of elements

Reply:
Yes, but these exact nodes represent the shape of the data, so in terms of shape nodes the rank of the data it represents is the number of elements in the shape data itself.